ml/src/config.py

```
@@ -183,6 +198,37 @@ def _validate(cfg: dict) -> None:
             "the only preprocessing contract the pipeline implements"
         )
 
+    for key, default in _OPTIONAL_PREPROCESSING_DEFAULTS.items():
+        pre.setdefault(key, default)
+
+    canonical = pre["canonical_mm_per_px"]
+    if not isinstance(canonical, (int, float)) or isinstance(canonical, bool):
+        raise ValueError("preprocessing.canonical_mm_per_px must be a number")
```

Ruff (0.16.3)

[warning] Prefer `TypeError` exception for invalid type

(TRY004)